File: Data Structures & Algorithms/partition-equal-subset-sum/submission-4.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        total = sum(nums)
        target = total // 2
        if total % 2:
            return False

        dp = [[False]*(target+1) for _ in range(n+1)]
        dp[0][0] = True
        for i in range(1, n+1):
            for j in range(target+1):
                dp[i][j] = dp[i-1][j]
                if j - nums[i-1] >= 0:
                    dp[i][j] = dp[i][j] or dp[i-1][j - nums[i-1]]
        
        return dp[n][target]

File: Data Structures & Algorithms/two-integer-sum/submission-4.rs
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        for i in 0..nums.len() {
            let key = target-nums[i];
            if map.contains_key(&key) {
                return vec![map[&key], i as i32];
            } else {
                map.insert(nums[i], i as i32);
            }
        }
        return vec![];
    }
}
